src/native/eventpipe/ep-buffer-manager.c

```diff
@@ -1209,12 +1213,33 @@ ep_buffer_manager_write_all_buffers_to_file_v4 (
 					// miscategorize it, but that seems unlikely.
 					uint32_t last_read_delta = last_read_sequence_number - thread_sequence_number;
 					if (0 < last_read_delta && last_read_delta < 0x80000000) {
-						ep_rt_thread_sequence_number_map_add_or_replace (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), session_state, last_read_sequence_number);
-						if (!exists)
+						if (exists) {
+							ep_rt_thread_sequence_number_map_remove (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), session_state);
+						} else {
 							ep_thread_addref (ep_thread_holder_get_thread (ep_thread_session_state_get_thread_holder_ref (session_state)));
+						}
+						ep_rt_thread_sequence_number_map_add (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), session_state, last_read_sequence_number);
+					}
+
+					// if a session_state was exhausted during this sequence point, mark it for deletion
+					if (ep_thread_session_state_get_buffer_list(session_state)->head_buffer == NULL) {
+
+						if (ep_rt_volatile_load_uint32_t_without_barrier(&(ep_thread_session_state_get_thread(session_state))->unregistered) > 0) {
```

Comment:
You should access unregistered filed through a getter since it is declared as part of ep-thread.c and not directly accessible to ep-buffer-manager (if compiled without inline getter/setter), so it is a "private" field.